Update analytics.js

Rafkev opened the pull request:
Great work bro, just added some enhancements to the code 

Improved User Identifier Generation: Rather than using a combination of Date.now() and Math.random() for generating a user identifier, it's better to use a more reliable method. We can use UUID (Universally Unique Identifier) for this purpose.

Ensure Organic Source Cookie Expiry: Adding an expiry date to the organic source cookie ensures that it doesn't persist indefinitely. This improves privacy and ensures the cookie doesn't accumulate unnecessarily.

Refactor Organic Source Detection: Refactor the getOrganicSource() function to make it more scalable and easier to maintain. Let's create a mapping of search engine domains to their respective names, making it simpler to add or modify search engines in the future.

Rafkev committed:
Update analytics.js

Great work bro, just added some enhancements to the code 

Improved User Identifier Generation: Rather than using a combination of Date.now() and Math.random() for generating a user identifier, it's better to use a more reliable method. We can use UUID (Universally Unique Identifier) for this purpose.

Ensure Organic Source Cookie Expiry: Adding an expiry date to the organic source cookie ensures that it doesn't persist indefinitely. This improves privacy and ensures the cookie doesn't accumulate unnecessarily.

Refactor Organic Source Detection: Refactor the getOrganicSource() function to make it more scalable and easier to maintain. Let's create a mapping of search engine domains to their respective names, making it simpler to add or modify search engines in the future.

Rafkev commented:
**More Info:**

**UUID for User Identifier:** Let's use the uuid library to generate UUIDs for user identification, which ensures uniqueness across different users and sessions.

**Expiry Date for Cookies:** When setting the organic source cookie, an expiry date is included. This ensures that the cookie is valid only for a certain duration (expiryDays), after which it will be automatically deleted by the browser.

**Refactored Organic Source Detection:** The getOrganicSource() function now uses a mapping of search engine domains to their names. This makes it easier to add or modify search engines in the future without changing the core logic of the function.

okeken reviewed (commented):
Thanks for your PRs!

The aim of using a very simple unique id implementation is to intentionally avoid build process that will come along with node modules to the project. 

There are some elegant approaches which will may consider.

1. Using crypto.randomUUID() (Modern Browsers) - this may need a polyfill for older browsers.
2. Custom UUID Generator
3. Date Based Unique ID ( a simpler was used here)
4. Create a standalone server to generate ID (this server will also be responsible for other activities such as database service, authentication etc)
Let me know which is best is suited for you.

All other code modifications looks neat,

Happy to  take another look.

Rafkev commented:
Okay, then I have updated the code with "Custom UUID" as you have advised. Let's consider creating a  function to generate UUIDs instead of using the uuid library. In this code, generateUUID() function generates a UUID similar to the uuidv4() function from the uuid library. I've replaced the usage of uuidv4() with this custom function. This way, UUIDs can be generated without importing an external library. Please have a look 


`// analytics.js

// Function to generate a unique user identifier using custom UUID generator
function generateUUID() {
  let dt = new Date().getTime();
  const uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
    const r = (dt + Math.random()*16)%16 | 0;
    dt = Math.floor(dt/16);
    return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
  });
  return uuid;
}

export function generateUserIdentifier() {
  let userIdentifier = localStorage.getItem("user_identifier");
  if (!userIdentifier) {
    userIdentifier = "user_" + generateUUID(); // Generate UUID using custom function
    localStorage.setItem("user_identifier", userIdentifier);
  }
  return userIdentifier;
}

// Function to set a cookie with the organic source information
export function setOrganicSourceCookie(organicSource, expiryDays = 7) {
  const expiryDate = new Date();
  expiryDate.setDate(expiryDate.getDate() + expiryDays); // Set expiry date
  const cookieValue = `organic_source=${organicSource}; expires=${expiryDate.toUTCString()}; path=/`;
  document.cookie = cookieValue;
}

// Function to get the organic source from the cookie
export function getOrganicSourceFromCookie() {
  const name = "organic_source=";
  const decodedCookie = decodeURIComponent(document.cookie);
  const cookieArray = decodedCookie.split(";");
  for (let i = 0; i < cookieArray.length; i++) {
    const cookie = cookieArray[i].trim();
    if (cookie.indexOf(name) === 0) {
      return cookie.substring(name.length, cookie.length);
    }
  }
  return "";
}

// Function to get the organic source
export function getOrganicSource() {
  const referrer = document.referrer;
  const searchEngines = {
    "google.com": "Google",
    "bing.com": "Bing",
    "yahoo.com": "Yahoo"
    // Add more search engines here if needed
  };

  let organicSource = "Other Organic Traffic";
  for (const domain in searchEngines) {
    if (referrer.includes(domain)) {
      organicSource = searchEngines[domain];
      break;
    }
  }

  return organicSource;
}
`
